src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+    function redeem(uint256 shares, address receiver, address owner) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 maxShares = maxRedeem(owner);
+        if (shares > maxShares) {
+            revert ExceededMaxRedeem(owner, shares, maxShares);
+        }
+
+        uint256 assets = previewRedeem(shares);
+
+        _withdraw(_msgSender(), receiver, owner, assets, shares);
+
+        return assets;
+    }
+
+    //// 4626-MAX ////
+
+    function getAssets() public view returns (address[] memory) {
+        return _getAssetStorage().list;
+    }
+
+    function getAsset(address asset_) public view returns (AssetParams memory) {
+        return _getAssetStorage().assets[asset_];
+    }
+
+    function getStrategies() public view returns (address[] memory) {
+        return _getStrategyStorage().list;
+    }
+
+    function getStrategy(address asset_) public view returns (StrategyParams memory) {
+        return _getStrategyStorage().strategies[asset_];
+    }
+
+    function previewDepositAsset(address asset_, uint256 assets_) public view returns (uint256) {
+        if (!_getAssetStorage().assets[asset_].active) revert InvalidAsset();
+        return _convertToShares(asset_, assets_, Math.Rounding.Floor);
+    }
+
+    function depositAsset(address asset_, uint256 assets_, address receiver) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+        if (!_getAssetStorage().assets[asset_].active) revert InvalidAsset();
+
+        uint256 shares = previewDepositAsset(asset_, assets_);
+        _deposit(asset_, _msgSender(), receiver, assets_, shares);
+
+        return shares;
+    }
+
+    function paused() public view returns (bool) {
+        return _getVaultStorage().paused;
+    }
+
+    function rateProvider() public view returns (address) {
+        return _getVaultStorage().rateProvider;
+    }
+
+    function bufferStrategy() public view returns (address) {
+        return _getVaultStorage().bufferStrategy;
+    }
+
+    function processAccounting() public {
+        AssetStorage storage assetStorage = _getAssetStorage();
+        StrategyStorage storage strategyStorage = _getStrategyStorage();
+        uint256 totalBaseBalance = 0;
+
+        for (uint256 i = 0; i < assetStorage.list.length; i++) {
+            address asset_ = assetStorage.list[i];
+            uint256 idleBalance = IERC20(asset_).balanceOf(address(this));
+            assetStorage.assets[asset_].idleBalance = idleBalance;
+            totalBaseBalance += _convertAssetToBase(asset_, idleBalance);
+        }
+
+        for (uint256 i = 0; i < strategyStorage.list.length; i++) {
+            address strategy = strategyStorage.list[i];
+            uint256 idleBalance = IERC20(strategy).balanceOf(address(this));
+            strategyStorage.strategies[strategy].idleBalance = idleBalance;
+            totalBaseBalance += _convertAssetToBase(strategy, idleBalance);
+        }
```

Comment:
we have to handle the asynchronous withdrawal positions for certain strategies with custom logic for ynETH and ynLSDe and potentially other asynchronous withdrawals strategies.

For ynETH and ynLSDe, you'd call 

withdrawalRequestsForOwner
https://github.com/yieldnest/yieldnest-protocol/blob/main/src/WithdrawalQueueManager.sol#L561
and sum the WithdrawalRequest.amount

For Karak you have Vault.getQueuedWithdrawal (no NFTs)
https://github.com/karak-network/v2-contracts/blob/main/src/Vault.sol#L258


In contrast, Kernel will be deployed with instant withdrawals.



https://github.com/Kelp-DAO/kernel-smart-contracts-public/blob/main/src/KernelVault.sol#L132

Going into this, there was the idea of packaging karak in an ynLSDk asset on its own. That i think may still makes sense since you have to not only handle staking/unstaking but also operator selection. With karak you create one vault per validator(?).

Regardless, It seems like here we'd need the custom module/modules functionality to handle withdrawal positions. In theory this way you could plug in other protocols as well as strategies (Eg. you can plug in WEETH as a tokenized strategy). 

For these modules we can either use inheritance (inherit this vault and just add the logic for each strategy) or additional Module contracts

Reply:
great feedback. I've refactored how the processor makes calls by adding a ProcessorStorage position that tracks a whitelisting of function signatures on contracts:

https://github.com/yieldnest/yieldnest-vault/blob/0ec0acf6d968612a3611127f67017d9c976d2306/src/Vault.sol#L219